pkg/odo/genericclioptions/context.go

```diff
@@ -0,0 +1,173 @@
+package genericclioptions
+
+import (
+	"fmt"
+	"github.com/golang/glog"
+	"github.com/redhat-developer/odo/pkg/application"
+	"github.com/redhat-developer/odo/pkg/component"
+	"github.com/redhat-developer/odo/pkg/occlient"
+	"github.com/redhat-developer/odo/pkg/odo/util"
+	"github.com/redhat-developer/odo/pkg/project"
+	"github.com/spf13/cobra"
+	"os"
+)
+
+// NewContext creates a new Context struct populated with the current state based on flags specified for the provided command
+func NewContext(command *cobra.Command) *Context {
+	return newContext(command, false)
+}
+
+// NewContextCreatingAppIfNeeded creates a new Context struct populated with the current state based on flags specified for the
+// provided command, creating the application if none already exists
+func NewContextCreatingAppIfNeeded(command *cobra.Command) *Context {
+	return newContext(command, true)
+}
+
+// Client returns an oc client configured for this command's options
+func Client(command *cobra.Command) *occlient.Client {
+	return client(command)
+}
+
+// ClientWithConnectionCheck returns an oc client configured for this command's options but forcing the connection check status
+// to the value of the provided bool, skipping it if true, checking the connection otherwise
+func ClientWithConnectionCheck(command *cobra.Command, skipConnectionCheck bool) *occlient.Client {
+	return client(command, skipConnectionCheck)
+}
+
+//
+func client(command *cobra.Command, shouldSkipConnectionCheck ...bool) *occlient.Client {
```

Comment:
why are we passing in multiple bools? shouldn't it just be `bool` instead of `...bool`?

Reply:
The idea here is to be able to determine whether a boolean is passed or not because we cannot rely on zero value for bool (which would be false). Since the method is generic, we need to determine whether or not we want to override the flag value so passing a single bool doesn't work. Using varargs allow a ternary state, making the bool optional so either it's present and its value should always override the flag or it is not, and we take the flag into account. It's not ideal but I preferred that solution as opposed to passing a pointer and check for nil (which didn't really make sense) or passing 2 bools where one would be ignored depending on the value of the first one.

Reply:
can we consider a Options like builder pattern?

```
type Option func(client *occlient.Client)

func SkipConnectionCheck(check bool) Option {
	return func(client *occlient.Client) {
		client.skipConnectionCheck = check
	}
}
```
Then the client function will work like this

```
func client(command *cobra.Command, options ...Option) *occlient.Client {
}

```
This can be useful if we want to add more options in the future.

Reply:
> can we consider a Options like builder pattern?
> 
> ```
> type Option func(client *occlient.Client)
> 
> func SkipConnectionCheck(check bool) Option {
> 	return func(client *occlient.Client) {
> 		client.skipConnectionCheck = check
> 	}
> }
> ```
> 
> Then the client function will work like this
> 
> ```
> func client(command *cobra.Command, options ...Option) *occlient.Client {
> }
> ```
> 
> This can be useful if we want to add more options in the future.

That's indeed an interesting idea that I hadn't considered. However, I'd rather not speculate on the future at this point and keep this PR as it is. We can revisit this idea at a later point if needed.

Reply:
@kadel Done: see #968